src/main/kotlin/app/revanced/patches/all/connectivity/wifi/spoof/patch/SpoofWifiPatch.kt

```diff
@@ -0,0 +1,232 @@
+            arrayOf("Landroid/net/NetworkRequest;", "Landroid/net/ConnectivityManager\$NetworkCallback;", "Landroid/os/Handler;"),
+            "V",
+        ),
+        RequestNetwork1(
+            "Landroid/net/ConnectivityManager;",
+            "requestNetwork",
+            arrayOf("Landroid/net/NetworkRequest;", "Landroid/net/ConnectivityManager\$NetworkCallback;"),
+            "V",
+        ),
+        RequestNetwork2(
+            "Landroid/net/ConnectivityManager;",
+            "requestNetwork",
+            arrayOf("Landroid/net/NetworkRequest;", "Landroid/net/ConnectivityManager\$NetworkCallback;", "I"),
+            "V",
+        ),
+        RequestNetwork3(
+            "Landroid/net/ConnectivityManager;",
+            "requestNetwork",
+            arrayOf("Landroid/net/NetworkRequest;", "Landroid/net/ConnectivityManager\$NetworkCallback;", "Landroid/os/Handler;"),
+            "V",
+        ),
+        RequestNetwork4(
+            "Landroid/net/ConnectivityManager;",
+            "requestNetwork",
+            arrayOf("Landroid/net/NetworkRequest;", "Landroid/app/PendingIntent;"),
+            "V",
+        ),
+        RequestNetwork5(
+            "Landroid/net/ConnectivityManager;",
+            "requestNetwork",
+            arrayOf("Landroid/net/NetworkRequest;", "Landroid/net/ConnectivityManager\$NetworkCallback;", "Landroid/os/Handler;", "I"),
+            "V",
+        );
+
+        fun replaceInstruction(method: MutableMethod, instruction: Instruction35c, instructionIndex: Int) {
+            val registers = arrayOf(instruction.registerC, instruction.registerD, instruction.registerE, instruction.registerF, instruction.registerG)
+            val argsNum = this.methodParams.size + 1 // + 1 for instance of definedClassName
+            if (argsNum > registers.size) {
+                // should never happen, but just to be sure (also for the future) a safety check
+                throw RuntimeException("not enough registers for ${this.definedClassName}#${this.methodName}: required $argsNum registers, but only got ${registers.size}")
+            }
+            val args = registers.take(argsNum).joinToString(separator = ", ") { reg -> "v${reg}" }
+            val replacementMethodDefinition = "${this.methodName}(${this.definedClassName}${this.methodParams.joinToString(separator = "")})${this.returnType}"
+            method.replaceInstruction(
+                instructionIndex,
+                "invoke-static { $args }, $INTEGRATIONS_CLASS_DESCRIPTOR->${replacementMethodDefinition}"
+            )
+        }
+
+        companion object {
+            fun fromMethodReference(methodReference: MethodReference) = values().firstOrNull { search ->
+                search.definedClassName == methodReference.definingClass
+                        && search.methodName == methodReference.name
+                        && methodReference.parameterTypes.toTypedArray().contentEquals(search.methodParams)
+            }
+        }
```

Comment:
This method is a duplicate of the method from `RemoveScreenshotRestrictionPatch` apart from the parameters. If that is not expected, this should be refactored to a single commonly used method.

Reply:
I’m not sure what would be a good way to refactor this, therefore I’d leave it as it is right now.

Reply:
A utility class inside the patches package can be created.

Reply:
Do you want me to only refactor the check if the methods match, or the entire `fromMethodReference` method?

Reply:
The entire method can be outsourced into a utility class which both patches can make use of.

Reply:
Could you tell me what the method definition of that utility function should look like? I still have no good idea.

Reply:
The method takes a MethodReference as a parameter and returns the first match out of your method call enum. Because you can not instantiate enum classes, you will have to rely on inheritencd in order to be able to utilize the utility method from both patches with different MethodCall types.

Reply:
Perhaps something like this?
```kotlin
interface IMethodCall {
    val definedClassName: String
    val methodName: String
    val methodParams: Array<String>
    val returnType: String
}

open class MethodCallResolver<E>(private val clazz: Class<E>) where E: IMethodCall, E: Enum<E> {
    fun fromMethodReference(methodReference: MethodReference) = clazz.enumConstants.firstOrNull { search ->
        search.definedClassName == methodReference.definingClass && search.methodName == methodReference.name
    }
}

enum class MethodCall(
    override val definedClassName: String,
    override val methodName: String,
    override val methodParams: Array<String>,
    override val returnType: String
): IMethodCall {
    SetFlags(
        "Landroid/view/Window;",
        "setFlags",
        arrayOf("I", "I"),
        "V",
    );

    companion object : MethodCallResolver<MethodCall>(MethodCall::class.java)
}
```

Reply:
You can generically constrain your method to the required types instead of having to instatiate a class MethodCallResolver for each ClassDef and instead of an enum, pass a list of your enums values mapped to the provided interface. This would allow getting rid of making reflective calls (which how it is right now can be improved by simply creating a field for the enum constants in your class MethodCallResolver). If you want I can take on this later this day if it's unclear.

Reply:
Like this?
```kotlin
interface IMethodCall {
    val definedClassName: String
    val methodName: String
    val methodParams: Array<String>
    val returnType: String
}

inline fun <reified E> fromMethodReference(methodReference: MethodReference)
        where E: Enum<E>, E: IMethodCall = enumValues<E>().firstOrNull { search ->
    search.definedClassName == methodReference.definingClass
            && search.methodName == methodReference.name
            && methodReference.parameterTypes.toTypedArray().contentEquals(search.methodParams)
}

enum class MethodCall(
    override val definedClassName: String,
    override val methodName: String,
    override val methodParams: Array<String>,
    override val returnType: String
): IMethodCall {
    SetFlags(
        "Landroid/view/Window;",
        "setFlags",
        arrayOf("I", "I"),
        "V",
    );
}
```

Reply:
That looks fine now.